### PR TITLE
client and server:put support auto lease

### LINKED ISSUE #12665: Feature discuss: we need auto-lease feature  for PUT
```
        lease, err := m.cliv3.Grant(ctx, ttl)
	if err != nil {
		return err
	}
	updateStatus := []clientv3.Op{clientv3.OpPut(statusKey, val, clientv3.WithLease(lease.ID))}
	tr, err := m.cliv3.Txn(ctx).
		If(clientv3.Compare(clientv3.Version(fmt.Sprintf(containerInfoKey, container.ID)), "!=", 0)).
		Then( 
			clientv3.OpTxn(
				[]clientv3.Cmp{clientv3.Compare(clientv3.Version(statusKey), "!=", 0)},
				[]clientv3.Op{clientv3.OpTxn( 
					[]clientv3.Cmp{clientv3.Compare(clientv3.Value(statusKey), "=", val)},
					[]clientv3.Op{clientv3.OpGet(statusKey)},
					updateStatus,                            
				)},
				updateStatus, // 
			),
		).Commit()
```
For the code above, we can see that if we want to update a key's lease that the value has been changed.
We must create a new lease for that key, and the lease may not be used if we don't need any update.

As keys increase, etcd will get more and more PUT requests with same value and eventually break down.

So for this two reason,we want  auto-lease feature.

**What auto-lease will do:**

1. auto generates a lease for  the given key with the ttl if the key not exist.
2. if key exists, but value not change. Prevent the write operation  and only renew the lease.
3. if value changes, renew the lease.
4. not functional in txn (because  etcd code structure not support lease ops in txn)

**Pros:**
1. convenient for users who want to do the similar job
2. reduce write ops for better performance

### PR DESCRIPTION
**Pull request by Tyraelone**
etcd server and client: PUT add a new feature about auto lease.  

Usage: put  --auto-lease --auto-lease-ttl 100 key value 
Description: create a lease with given key if there is no lease, or renew lease if it exists 
Why:I want to let ETCD put method to support auto-renew lease when value has not changed, for reducing write ops. 
Problem:ETCD can not do lease ops in a transaction
Fixes #12665

### CODE REVIEW

**Comment by Tyraelone** [comment]
@jingyih jin @spzala  any idea about this?   We really want a feature like this.

**Comment by CMGS** [comment]
no one care about this PR? I think it necessary to support update lease in one txn. Good job tyarelone.

**Review by tangcong — CHANGES_REQUESTED** [review]
I take a brief look, there are the following suggestions
1. split commit by feature
2. add test cases
3. add trace for auto lease
4. format indentation

**Comment by Tyraelone** [comment]
> I take a brief look, there are the following suggestions
> 
> 1. split commit by feature
> 2. add test cases
> 3. add trace for auto lease
> 4. format indentation

Thanks, let me fill these things

**Comment by Tyraelone** [comment]
@tangcong tan
added test cases for auto lease

**Comment by Tyraelone** [comment]
@jingyih @tangcong @spzala

**Comment by CMGS** [comment]
no update? I'm wondering who in charge of this project? no discussion no one care no response? is it the way you guys run this project and community?

File: api/etcdserverpb/rpc.proto [hunk 1]
```diff
@@ -14,29 +14,29 @@ option (gogoproto.unmarshaler_all) = true;
 service KV {
   // Range gets the keys in the range from the key-value store.
   rpc Range(RangeRequest) returns (RangeResponse) {
-      option (google.api.http) = {
```
Comment (tangcong) [inline api/etcdserverpb/rpc.proto:17]: ?

**Review by tangcong — CHANGES_REQUESTED** [review]
@Tyraelone @CMGS sorry I have been very busy recently. I take a look this PR again and found that there are still some problems that have not been solved. please also take a look. @ptabor

File: api/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -497,6 +497,9 @@ message PutRequest {
   // If ignore_lease is set, etcd updates the key using its current lease.
   // Returns an error if the key does not exist.
   bool ignore_lease = 6;
+
+  bool auto_lease = 7;
+  int64 ttl = 8;
```
[the remark this answers is not in the document]
Comment (Tyraelone) [reply api/etcdserverpb/rpc.proto:502]: ok,i will fix it @tangcong

**Comment by ptabor** [comment]
@gyuho @jpbetz 
Its pretty significant change as it's affect public API surface. Such proposal IMHO should be discussed as Issue / doc first to decide about the desired semantic. I think we should go back to that phase. 

Few quick thoughts: 
   
   1. The PR promotes lease-per-object model that for high throughput cases was considered inefficient. Please see related discussion in Kubernetes:  https://github.com/kubernetes/kubernetes/pull/98257. 
       If we wanted to integrate auto-leasing future in etcd, maybe it should perform similar batching model. 
       
   2. What's the main purpose of the feature ?  
            a) More convenient API ? 
            b) Better performance thanks to less objects in the wire format ? 
            c) sth else
      If that's a)  -> the design should consider option when the code is contributed as part of client/v3 lib (or wrapper on top of the library).

**Comment by Tyraelone** [comment]
> @gyuho @jpbetz
> Its pretty significant change as it's affect public API surface. Such proposal IMHO should be discussed as Issue / doc first to decide about the desired semantic. I think we should go back to that phase.
> 
> Few quick thoughts:
> 
> 1. The PR promotes lease-per-object model that for high throughput cases was considered inefficient. Please see related discussion in Kubernetes:  [kubernetes/kubernetes#98257](https://github.com/kubernetes/kubernetes/pull/98257).
>    If we wanted to integrate auto-leasing future in etcd, maybe it should perform similar batching model.
> 2. What's the main purpose of the feature ?
>    a) More convenient API ?
>    b) Better performance thanks to less objects in the wire format ?
>    c) sth else
>    If that's a)  -> the design should consider option when the code is contributed as part of client/v3 lib (or wrapper on top of the library).
Sorry for my poor English first.
1: we manage container health status with etcd. After a health check, we  only want to renew the lease if status not change to avoid more write ops.
2 a)  yes, convenient 
  b)  the main reason of this PR is that  we don't want to really do PUT ops if data not change

File: api/etcdserverpb/rpc.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tyraelone) [reply api/etcdserverpb/rpc.proto:504]: @tangcong  the swag is auto generated, I need to push it too?

[the remark this answers is not in the document]
Comment (Tyraelone) [reply api/etcdserverpb/rpc.proto:504]: @tangcong tang fixed.

**Pull request closed without merging**